[Adding] ability to revoke user by nkey only

matthiashanel opened the pull request:
Fixes #302

Signed-off-by: Matthias Hanel <mh@synadia.com>

aricart commented on cmd/revokeuser.go:
```diff
@@ -40,6 +40,7 @@ func createRevokeUserCmd() *cobra.Command {
 	cmd.Flags().StringVarP(&params.user, "name", "n", "", "user name")
 	cmd.Flags().IntVarP(&params.at, "at", "", 0, "revokes all user credentials created before a Unix timestamp ('0' is treated as now)")
+	cmd.Flags().StringVarP(&params.userPubKey, "public-user-key", "u", "", "public user key (use when user was generated outside of nsc)")
 
```
--user-public-key is prob the right flag name

aricart commented on cmd/revokeuser.go:
```diff
@@ -110,33 +111,38 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
-	if p.user == "" {
-		return fmt.Errorf("user is required")
+	if p.userPubKey == "" && p.user == "" {
+		return fmt.Errorf("user or public-user-key are required")
```
same here

aricart commented on cmd/revokeuser.go:
```diff
@@ -110,33 +111,38 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
+	if p.userPubKey == "" && p.user == "" {
+		return fmt.Errorf("user or public-user-key are required")
+	} else if p.userPubKey != "" && p.user != "" {
+		return fmt.Errorf("user and public-user-key are mutually exclusive")
```
and here.

aricart commented on cmd/revokeuser.go:
```diff
@@ -110,33 +111,38 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
-		return fmt.Errorf("user is required")
+		p.userPubKey = userClaim.Subject
+	} else if !nkeys.IsValidPublicUserKey(p.userPubKey) {
+		return fmt.Errorf("public-user-key %q is not a valid public user nkey", p.userPubKey)
```
public-user-key

aricart commented on cmd/revokeuser.go:
```diff
@@ -110,33 +111,38 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
-	if err != nil || userClaim == nil {
-		return fmt.Errorf("user is required")
+		p.userPubKey = userClaim.Subject
+	} else if !nkeys.IsValidPublicUserKey(p.userPubKey) {
```
Key validation should happen on the validate, that way the interactive doesn't have to validate as well

matthiashanel replied:
fixed

matthiashanel commented:
Added support to clear revocations by nkey.

Changed the interactive mode for clear to only show revoked nkeys (vs. all users)
When available the user name is presented too.

Also aliased commands to work with -

```
 16:22:52  ✘ 1  ~/repos/nsc  revoke-by-pub *13 ?1  nsc revocations delete-user -u UCLMBZ5CBDRDG2TAYOJK23U7IGKPTTW7DTWNOP4TUW4PAB3GRUSKXG3N
[ OK ] cleared user revocation for account UCLMBZ5CBDRDG2TAYOJK23U7IGKPTTW7DTWNOP4TUW4PAB3GRUSKXG3N
Cleared revocation of user  with public key UCLMBZ5CBDRDG2TAYOJK23U7IGKPTTW7DTWNOP4TUW4PAB3GRUSKXG3N
 16:23:02  ~/repos/nsc  revoke-by-pub *13 ?1  nsc revocations list_users
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│                                 Revoked Users for testme                                 │
├──────────────────────────────────────────────────────────┬───────────────────────────────┤
│ Public Key                                               │ Revoke Credentials Before     │
├──────────────────────────────────────────────────────────┼───────────────────────────────┤
│ UAZZM3O3EP2EGGQFCBZPIXEKGNXTHT6USNM3BXMQFVX5M65TQCNN3H24 │ Fri, 02 Oct 2020 15:18:20 EDT │
╰──────────────────────────────────────────────────────────┴───────────────────────────────╯
 16:23:16  ~/repos/nsc  revoke-by-pub *13 ?1  nsc revocations delete-user -u UAZZM3O3EP2EGGQFCBZPIXEKGNXTHT6USNM3BXMQFVX5M65TQCNN3H24
[ OK ] cleared user revocation for account UAZZM3O3EP2EGGQFCBZPIXEKGNXTHT6USNM3BXMQFVX5M65TQCNN3H24
Cleared revocation of user  with public key UAZZM3O3EP2EGGQFCBZPIXEKGNXTHT6USNM3BXMQFVX5M65TQCNN3H24
 16:23:18  ~/repos/nsc  revoke-by-pub *13 ?1  nsc revocations list_users
╭────────────────────────────────────────╮
│        Revoked Users for testme        │
├────────────┬───────────────────────────┤
│ Public Key │ Revoke Credentials Before │
├────────────┼───────────────────────────┤
╰────────────┴───────────────────────────╯
 16:23:19  ~/repos/nsc  revoke-by-pub *13 ?1 
```

aricart commented on cmd/revokeclearuser.go:
```diff
@@ -74,11 +79,43 @@ func (p *ClearRevokeUserParams) PreInteractive(ctx ActionCtx) error {
+
+	st := ctx.StoreCtx().Store
+	accClaim, err := st.ReadAccountClaim(p.AccountContextParams.Name)
+	if err != nil {
```
if err != nil || len(accClaim.Revocations) == 0 {
   return err
}

matthiashanel replied:
fixed

aricart commented on cmd/revokeclearuser.go:
```diff
@@ -74,11 +79,43 @@ func (p *ClearRevokeUserParams) PreInteractive(ctx ActionCtx) error {
+		return nil
+	}
+	keyToUserName := map[string]string{}
+	if users, err := st.ListEntries(store.Accounts, p.AccountContextParams.Name, store.Users); err != nil {
```
to me this reads better:
users, err := ...
if err != nil {
  return err
}

// no else required

aricart commented on cmd/revokeclearuser.go:
```diff
@@ -93,22 +130,26 @@ func (p *ClearRevokeUserParams) Load(ctx ActionCtx) error {
-		return fmt.Errorf("user is required")
+	if p.userPubKey == "" && p.user == "" {
+		return fmt.Errorf("user or user-public-key are required")
+	} else if p.userPubKey != "" && p.user != "" {
```
there's no else...

aricart commented on cmd/revokeuser.go:
```diff
@@ -75,20 +77,24 @@ func (p *RevokeUserParams) canParse(s string) error {
-		p.user, err = ctx.StoreCtx().PickUser(p.AccountContextParams.Name)
-		if err != nil {
-			return err
+	if byName, err := cli.Confirm("Do you want to revoke a user (known to nsc) by name?", true); err != nil {
```
byName, err := ...
if err != nil { 
  return...
}

aricart commented on cmd/revokeuser.go:
```diff
@@ -110,22 +117,25 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
 		n := ctx.StoreCtx().DefaultUser(p.AccountContextParams.Name)
 		if n != nil {
 			p.user = *n
 		}
```
if n == nil {
   fmt.Errorf("some error)
}
p.user = *n

aricart commented on cmd/revokeuser.go:
```diff
@@ -110,22 +117,25 @@ func (p *RevokeUserParams) Load(ctx ActionCtx) error {
-		return fmt.Errorf("user is required")
+	if p.userPubKey == "" && p.user == "" {
+		return fmt.Errorf("user or user-public-key are required")
+	} else if p.userPubKey != "" && p.user != "" {
```
this won't happen with the above...

matthiashanel replied:
it will, if the above does not set a user

aricart reviewed (approved):
the if/if else/else type stuff gets really dense to look at - I think you should simplify it.
